CIP-ledger-evolution/README.md

```diff
@@ -0,0 +1,302 @@
+[small step operational semantics](STS.md).
+We refer to this framework as the *Small Step Semantics for Cardano*, or the *STS* for short.
+An understanding of the existing STS specifications for the
+[existing ledger eras](https://github.com/input-output-hk/cardano-ledger#cardano-ledger)
+is often required to fully understand the implications of changes to the ledger
+(though an understanding of the Haskell implementation is a fair substitute).
+
+The STS framework leaves both cryptographic primitives and the serialization format abstract.
+These specifications need to be complete enough to realize a full implementation of the ledger
+given the cryptographic primitives and serialization format.
+The cryptographic primitives are described as appendices to the STS specifications,
+and the serialization format is given as a
+[CDDL file](https://www.rfc-editor.org/rfc/rfc8610).
+There SHOULD be one STS specification per ledger era
+(the Allegra ledger era is currently combined with the Mary era).
+
+From the Byron to the Babbage ledger eras, the STS frameworks were written in $\LaTeX$.
+Starting in the Conway ledger era, literate Agda will be used.
+During the transition from $\LaTeX$ to literate Agda, we will take advantage
+of the ability to substitute $\LaTeX$ in place of Agda code when needed for expedience.
+With time, the Agda specification will gloriously blur the line between specification
+and reference implementation,
+just as the type system upon which Agda is built blurs the line between proof and program.
+
+### Ledger eras
+
+A ledger era is a collection of features added to the ledger which are introduced at a hard fork.
+The existing ledger eras, with a very simplistic description is given below.
+
+|name|new features|
+| --- | --- |
+|Byron|initial UTxO ledger|
+|Shelley|decentralized block production, stake delegation|
+|Allegra|timelock scripts|
+|Mary|multi-assets|
+|Alonzo|Plutus scripts|
+|Babbage|improved Plutus script contexts|
+|Conway|constitutional committee replacement, SPO voting on hard forks|
+
+
+### Ledger state
+
+The ledger state is the source of truth from which the network derives meaning.
+For performance reasons, we require it to be the minimal state needed
+to derive this value and validate blocks.
+
+There is currently one exception to the minimality, which will be deprecated in the days ahead,
+namely the stake pool ranking information which is used by Daedalus.
+
+#### Ledger events
+
+Some provenance about the ledger calculations is provided by ledger events.
+These events come with zero cost to a running node if not used and are not stored in the ledger state.
+
+### Soft forks and Hard forks
+
+We make the following definitions:
+
+**Hard fork** - A *hard fork* is a change to the protocol (not restricted to the ledger)
+resulting in a single new (CBOR) block being valid.
+
+**Soft fork** - A *soft fork* is a change to the protocol (not restricted to the ledger)
+resulting in fewer blocks being valid.
```

Comment:
I think my definition are correct, let me see if I can convince you.

Let `A` bet the set of valid block using the "current software". Let `B` be the set of valid blocks for some proposed change to the software.

I am defining a hard fork to be the case when `A \ B` is non-empty. Which I described in english as:

> A *hard fork* is a change to the protocol (not restricted to the ledger)
resulting in a single new (CBOR) block being valid.

So if we are not in the hard fork case, then `B` must be a subset of `A`. We then define a soft fork to be the situation when `B` is a proper-subset of `A`, which I've described in english as:

> A *soft fork* is a change to the protocol (not restricted to the ledger)
resulting in fewer blocks being valid.

Moreover, I think that my definitions are in line with the link you provided. They essentially define a fork as a change to the definition of a valid block, with the difference between hard and soft being whether or not a _validating_ only node (ie not block producing) would need to upgrade their software:

> [soft fork section] The old blockchain will keep accepting blocks from the new advanced blockchain platform since the fork is a forward-compatible alteration, although the regulations have been modified due to the new upgrade.

Reply:
The fact you need to write 15 lines of explanation is interesting, it suggests that either I don't have much more knowledge than a layperson (and thus should probably not be reviewing this document), or that the definition is not self-standing 🤔

Reply:
I'll assume it's me being confusing!!

Did you like my "venn diagram" explanation above?

Reply:
Sorry Arnaud, I didn't mean to be mean there :| FWIW the definition Jared gave is the one I've always seen for hard-/soft-forks and I didn't think it was unusual amongst people who are familiar with blockchains. Maybe it's a Duncan favourite or something.

Reply:
FWIW, I found all the only line definitions of hard fork and soft fork too vague (and thus confusing) until I finally saw the one above. It's very likely a Duncan original. :)

Reply:
@michaelpj no offence taken :) I am still having a hard time to have an intuition about what they mean, perhaps because they are too mathematical and abstract. 
To me, hard/soft forks relate 3 different things: The blocks forming the chain, the protocol, and the nodes running the software implementing the protocol and forming the chain:
* A hard fork is a change to the protocol which is backward incompatible: Software not implementing the new protocol simply won't be able to take part in the protocol anymore, and will be "stuck in time" unable to validate/produce more blocks
* A soft fork is a change of the protocol which is backward compatible: Software not implementing the new protocol will still be able to take part in it and will validate/produce blocks as before
I may be wrong, as always.

Reply:
> Software not implementing the new protocol simply won't be able to take part in the protocol anymore

ie `B` (valid blocks in the new software) admits blocks that `A` (valid blocks in the old software) does not. ie `B \ A` is non-empty.

> Software not implementing the new protocol will still be able to take part in it

ie `B` is a proper subset of `A`.

> and will validate/produce blocks as before

In the soft fork case, no, you _must_ upgrade if you want to _produce_ blocks.